mypy/subtypes.py

```diff
@@ -302,6 +331,237 @@ def visit_type_type(self, left: TypeType) -> bool:
 
 
+@contextmanager
+def pop_on_exit(stack: List[Tuple[Instance, Instance]],
+                left: Instance, right: Instance) -> Iterator[None]:
+    stack.append((left, right))
+    yield
+    stack.pop()
+
+
+def is_protocol_implementation(left: Instance, right: Instance, allow_any: bool = True) -> bool:
+    """Check whether 'left' implements the protocol 'right'. If 'allow_any' is False, then
+    check for a proper subtype. Treat recursive protocols by using the 'assuming'
+    structural subtype matrix (in sparse representation, i.e. as a list of pairs
+    (subtype, supertype)), see also comment in nodes.TypeInfo. When we enter a check for classes
+    (A, P), defined as following::
+
+      class P(Protocol):
+          def f(self) -> P: ...
+      class A:
+          def f(self) -> A: ...
+
+    this results in A being a subtype of P without infinite recursion. On every false result,
+    we pop the assumption, thus avoiding an infinite recursion as well.
+    """
+    assert right.type.is_protocol
+    assuming = right.type.assuming if allow_any else right.type.assuming_proper
+    for (l, r) in reversed(assuming):
+        if sametypes.is_same_type(l, left) and sametypes.is_same_type(r, right):
+            return True
+    with pop_on_exit(assuming, left, right):
+        for member in right.type.protocol_members:
+            # nominal subtyping currently ignores '__init__' and '__new__' signatures
+            if member in ('__init__', '__new__'):
+                continue
+            # The third argiment below indicates to what self type is bound.
+            # We always bind self to the subtype. (Similarly to nominal types).
+            supertype = find_member(member, right, left)
+            assert supertype is not None
+            subtype = find_member(member, left, left)
+            # Useful for debugging:
+            # print(member, 'of', left, 'has type', subtype)
+            # print(member, 'of', right, 'has type', supertype)
+            if not subtype:
+                return False
+            if allow_any:
+                # nominal check currently ignores arg names
+                is_compat = is_subtype(subtype, supertype, ignore_pos_arg_names=True)
+            else:
+                is_compat = is_proper_subtype(subtype, supertype)
+            if not is_compat:
+                return False
+            if isinstance(subtype, NoneTyp) and member.startswith('__') and member.endswith('__'):
+                # We want __hash__ = None idiom to work even without --strict-optional
+                return False
+            subflags = get_member_flags(member, left.type)
+            superflags = get_member_flags(member, right.type)
+            if IS_SETTABLE in superflags:
+                # Check opposite direction for settable attributes.
+                if not is_subtype(supertype, subtype):
+                    return False
+            if (IS_CLASSVAR in subflags and IS_CLASSVAR not in superflags or
```

Comment:
The check could be simplified to `(IS_CLASS_VAR in subflags) != (IS_CLASS_VAR in superflags)`.